Source/Core/Core/IOS/USB/Emulated/Infinity.h

```diff
@@ -0,0 +1,117 @@
+#include <queue>
+#include <vector>
+
+#include "Common/CommonTypes.h"
+#include "Common/IOFile.h"
+#include "Core/IOS/USB/Common.h"
+
+namespace IOS::HLE::USB
+{
+// Enum type, used when generating blank toys
+enum InfinityFigureType
+{
+  CHARACTER = 1,
+  PLAYSET = 2,
+  ABILITY_A = 3,
+  ABILITY_B = 4,
+  ITEM = 5
+};
+struct InfinityFigureInfo final
+{
+  InfinityFigureType type;
+  u16 figure_number;
+};
+class InfinityUSB final : public Device
+{
+public:
+  InfinityUSB(Kernel& ios, const std::string& device_name);
+  ~InfinityUSB();
+  DeviceDescriptor GetDeviceDescriptor() const override;
+  std::vector<ConfigDescriptor> GetConfigurations() const override;
+  std::vector<InterfaceDescriptor> GetInterfaces(u8 config) const override;
+  std::vector<EndpointDescriptor> GetEndpoints(u8 config, u8 interface, u8 alt) const override;
+  bool Attach() override;
+  bool AttachAndChangeInterface(u8 interface) override;
+  int CancelTransfer(u8 endpoint) override;
+  int ChangeInterface(u8 interface) override;
+  int GetNumberOfAltSettings(u8 interface) override;
+  int SetAltSetting(u8 alt_setting) override;
+  int SubmitTransfer(std::unique_ptr<CtrlMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<BulkMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IntrMessage> message) override;
+  int SubmitTransfer(std::unique_ptr<IsoMessage> message) override;
+  void ScheduleTransfer(std::unique_ptr<TransferCommand> command, const std::array<u8, 32>& data,
+                        s32 expected_count, u64 expected_time_us);
+
+private:
+  Kernel& m_ios;
+  u16 m_vid = 0;
+  u16 m_pid = 0;
+  u8 m_active_interface = 0;
+  bool m_device_attached = false;
+  DeviceDescriptor m_device_descriptor;
+  std::vector<ConfigDescriptor> m_config_descriptor;
+  std::vector<InterfaceDescriptor> m_interface_descriptor;
+  std::vector<EndpointDescriptor> m_endpoint_descriptor;
+  std::queue<std::array<u8, 32>> m_queries;
+  std::queue<std::unique_ptr<IntrMessage>> m_response_list;
+};
+
+struct InfinityFigure final
+{
+  File::IOFile inf_file;
+  std::array<u8, 0x14 * 0x10> data{};
+  bool present = false;
+  u8 order_added = 255;
+  void Save();
+};
+
+class InfinityBase final
+{
+public:
+  bool HasFigureBeenAddedRemoved() const;
+  std::array<u8, 32> GetAddedRemovedResponse();
+  void GetBlankResponse(u8 sequence, u8* reply_buf);
+  void GetPresentFigures(u8 sequence, u8* reply_buf);
+  void GetFigureIdentifier(u8 fig_num, u8 sequence, u8* reply_buf);
+  void QueryBlock(u8 fig_num, u8 block, u8* reply_buf, u8 sequence);
+  void WriteBlock(u8 fig_num, u8 block, const u8* to_write_buf, u8* reply_buf, u8 sequence);
+  void RemoveFigure(u8 position);
+  std::string LoadFigure(u8* buf, File::IOFile in_file, u8 position);
+  bool CreateFigure(const std::string& file_path, u16 character);
+  const std::map<const char*, const InfinityFigureInfo> GetFigureList();
+  std::string FindFigure(u16 character) const;
+  void GenerateSeed(u32 seed);
```

Comment:
These functions need to remain public so that they can be called from the USB Transfer methods and the Qt window, via the Core::System reference to the infinity base